File: 91-decode-ways/decode-ways.py
class Solution(object):
    def __init__(self):
        cache={}

    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        self.cache={}
        return self.decode(s)
    def decode(self,s):
        if s in self.cache:
            return self.cache[s]
        if s=="":
            return 1
        if s[0]=="0":
            return 0
        if len(s)==1:
            return 1
        res= self.decode(s[1:])
        if s[0] =="1" or (s[0]=='2' and s[1] in "0123456"):
            res+=self.decode(s[2:])
        self.cache[s]=res
        return res
        

File: 91-decode-ways/decode-ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        vector<int> cache(s.length() + 1, 0);
        cache[s.length()] = 1;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s[i] == '0') {
                cache[i] = 0;
                continue;
            }
            int count = 0;
            if (i != s.length() - 1 &&
                (s[i] == '1' || (s[i] == '2' && s[i + 1] <= '6')))
                count += cache[i + 2];
            count += cache[i + 1];
            cache[i] = count;
        }
        return cache[0];
    }
};